File: typescript/strain/strain.ts
function findersKeepers<T>(collection: Array<T>, predicate: (x: T) => Boolean, keepOnPredicateMatch: Boolean ): Array<T> {
    let keepers: Array<T> = [];

    collection.forEach(function(item: T) {
        if (predicate(item) == keepOnPredicateMatch) {
            keepers.push(item)
        }
    });

    return keepers;
}

export function keep<T>( collection: Array<T>, predicate: (x: T) => Boolean ): Array<T> {
    return findersKeepers(collection, predicate, true);
}

export function discard<T>( collection: Array<any>, predicate: (x: T) => Boolean ): Array<T> {
    return findersKeepers(collection, predicate, false);
}

File: solutions/csharp/strain/2/Strain.cs
public static class Strain
{
    public static IEnumerable<T> Keep<T>(this IEnumerable<T> collection, Func<T, bool> predicate) => Evaluate(collection, predicate, true);
    public static IEnumerable<T> Discard<T>(this IEnumerable<T> collection, Func<T, bool> predicate) => Evaluate(collection, predicate, false);

    private static List<T> Evaluate<T>(IEnumerable<T> collection, Func<T, bool> predicate, bool condition)
    {
        var results = new List<T>();

        foreach (var item in collection)
        {
            if (predicate(item) == condition)
            {
                results.Add(item);
            }
        }

        return results;
    }

}